aws/resource_aws_api_gateway_vpc_link.go

```diff
@@ -0,0 +1,200 @@
+package aws
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/apigateway"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsApiGatewayVpcLink() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsApiGatewayVpcLinkCreate,
+		Read:   resourceAwsApiGatewayVpcLinkRead,
+		Update: resourceAwsApiGatewayVpcLinkUpdate,
+		Delete: resourceAwsApiGatewayVpcLinkDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"target_arn": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceAwsApiGatewayVpcLinkCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).apigateway
+
+	input := &apigateway.CreateVpcLinkInput{
+		Name:       aws.String(d.Get("name").(string)),
+		TargetArns: []*string{aws.String(d.Get("target_arn").(string))},
+	}
+	if v, ok := d.GetOk("description"); ok {
+		input.Description = aws.String(v.(string))
+	}
+
+	resp, err := conn.CreateVpcLink(input)
+	if err != nil {
+		return err
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{apigateway.VpcLinkStatusPending},
+		Target:     []string{apigateway.VpcLinkStatusAvailable},
+		Refresh:    apigatewayVpcLinkRefreshStatusFunc(conn, *resp.Id),
+		Timeout:    8 * time.Minute,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("[WARN] Error waiting for APIGateway Vpc Link status to be \"%s\": %s", apigateway.VpcLinkStatusAvailable, err)
+	}
+
+	d.SetId(*resp.Id)
```

Comment:
We usually set the ID as soon as we can after receiving it from the API, _before_ making any further requests so that it's recorded in the state in case any future API requests fail for any reason. Do you mind moving it above the `StateChangeConf` block?